geps/gep-1731/index.md

```diff
@@ -0,0 +1,436 @@
+    //
+    // When this field is unspecified, the number of times to attempt to retry
+    // a backend request is implementation-specific.
+    //
+    // Support: Extended
+    //
+    // +optional
+    Attempts *Int `json:"attempts,omitempty"`
+
+    // Backoff specifies the minimum duration a Gateway should wait between
+    // retry attempts and is represented in Gateway API Duration formatting.
+    //
+    // For example, setting the `rules[].retry.backoff` field to the value
+    // `100ms` will cause a backend request to first be retried approximately
+    // 100 milliseconds after timing out or reciveing a response code configured
+    // to be retryable.
+    //
+    // An implementation MAY use an exponential or alternative backoff strategy
+    // for subsequent retry attempts, MAY cap the maximum backoff duration to
+    // some amount greater than the specified minimum, and MAY add arbitrary
+    // jitter to stagger requests, as long as unsuccessful backend requests are
+    // not retried before the configured minimum duration.
+    //
+    // If a Request timeout (`rules[].timeouts.request`) is configured on the
+    // route, the entire duration of the initial request and any retry attempts
+    // MUST not exceed the Request timeout duration. If any retry attempts are
+    // still in progress when the Request timeout duration has been reached,
+    // these SHOULD be cancelled if possible and the Gateway MUST immediately
+    // return a timeout error.
+    //
+    // If a BackendRequest timeout (`rules[].timeouts.backendRequest`) is
+    // configured on the route, any retry attempts which reach the configured
+    // BackendRequest timeout duration without a response SHOULD be cancelled if
+    // possible and the Gateway should wait for at least the specified backoff
+    // duration before attempting to retry the backend request again.
+    //
+    // If a BackendRequest timeout is _not_ configured on the route, retry
+    // attempts MAY time out after an implementation default duration, or MAY
+    // remain pending until a configured Request timeout or implementation
+    // default duration for total request time is reached.
+    //
+    // When this field is unspecified, the time to wait between retry attempts
+    // is implementation-specific.
+    //
+    // Support: Extended
+    //
+    // +optional
+    Backoff *Duration `json:"backoff,omitempty"`
+}
+
+// HTTPRouteRetryStatusCode defines an HTTP response status code for
+// which a backend request should be retried.
+//
+// Implementations MUST support the following status codes as retryable:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
+//
+// Implementations SHOULD support the `"5xx"` shorthand for matching status
+// codes in the 500-599 range. For implementations supporting `"5xx"`, this
+// shorthand MUST match the following status codes:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
```

Comment:
I think if we're going to define 5xx so narrowly (all of the status codes we require to be supported combined), we may want to just make this a standard value that also must be supported.

Reply:
An alternative would be to _require_ that 5xx matched all codes between 500-599 which would be less likely to surprise users but also less portable. The more I think about this, the more I think that we should require 5xx to support the full range of codes and consider it extended. (I know this somewhat conflicts with some of my earlier ideas).

Reply:
Hmm, agreed. I think that I would be pretty surprised if `5xx` did not mean "any code in the 500 range", rather than specific ones in a list.

Reply:
We could also consider adding something like Envoy's `gateway-error` which includes 502, 503 and 504 (notably lacking 500, which I think we would want to include).

Reply:
How about:

- `gateway-error`: 502, 503, 504
- `5xx`: 500-599
- if you want 500, 502, 503, 504 you use `[ "500", "gateway-error" ]`

??

I could also imagine `common-5xx` or `retriable-5xx` for 500, 502, 503, 504, but I'm not really sure that's _better_, just different.

Reply:
I'm starting to lean toward just removing `5xx` if it's most logical for it to be 500-599 because the existence of 501 implies the need for an additional `excludeCode` stanza only to support 5xx which is a bit wonky and verbose.

Reply:
It's probably fine to take out `5xx` for now, and add _something_ back in once we get feedback about what people actually want.

Reply:
Removed 5xx shorthand and switched HTTPRouteRetryStatusCode from `string` to `int` in 04e31dc33788ffef145ede61ae18e17b63720a1e